examples/hf_serialization.py

```diff
@@ -242,37 +325,36 @@ def hf_main(args):
     logger.info("GPU: " + utils.get_gpu_name())
     logger.info("PYTHON USED RAM: " + utils.get_mem_usage())
 
-    serialize_model(model, model_config, output_prefix)
-
-    tokenizer = AutoTokenizer.from_pretrained(args.input_directory)
-    tokenizer.save_pretrained(output_prefix)
+    serialize_model(model, model_config, output_prefix, None, args.force)
 
     if args.validate:
-        del model, model_config
+        # Not sure if this part is needed as, although I doubt it,
+        # I'm not absolutely certain whether or not TensorSerializer
+        # changes the model's state in any way
+        del model
         gc.collect()
         device = utils.get_device()
+        model = AutoModelForCausalLM.from_pretrained(
+            args.input_directory,
+            config=model_config,
+            torch_dtype=dtype,
+            low_cpu_mem_usage=True,
+        )
+        model.to(device)
         logger.info("Validating serialization")
-        model = load_model(
+        model2 = load_model(
             output_prefix,
             AutoModelForCausalLM,
             AutoConfig,
             None,
             device,
             dtype,
         ).eval()
-        # test generation
-        eos = tokenizer.eos_token_id
-        input_ids = tokenizer.encode(
-            "¡Hola! Encantado de conocerte. hoy voy a", return_tensors="pt"
-        ).to(device)
-        with torch.no_grad():
-            output = model.generate(
-                input_ids, max_new_tokens=50, do_sample=True, pad_token_id=eos
-            )
-        logger.info(
-            "Test Output:"
-            f" {tokenizer.decode(output[0], skip_special_tokens=True)}"
-        )
+        # Comparing model parameters
+        logger.info("Testing for sameness of model parameters")
+        for name, param in model.named_parameters():
+            param2 = model2.state_dict()[name]
+            assert torch.allclose(param, param2, atol=1e-3)
```

Comment:
1. This check's logic will miss if the serialized model somehow got extra fields (like, say, duplicates) that weren't present in the original
2. That `logger.info` could probably be `logger.debug`
3. Calling `.state_dict()` is slow, so it's better to put that before the loop
4. The values should be exactly equal. If they slightly off enough to need `torch.allclose` rather than `torch.equal`, that's bad
5. `assert` isn't good for testing correctness because `assert` statements can be skipped
```suggestion
        tensorizer_model = load_model(
            output_prefix,
            AutoModelForCausalLM,
            AutoConfig,
            None,
            device,
            dtype,
        ).eval()
        # Comparing model parameters
        logger.debug("Testing for sameness of model parameters")
        original_sd = model.state_dict()
        tensorizer_sd = tensorizer_model.state_dict()
        if not tensorizer_sd.keys() == original_sd.keys():
            raise RuntimeError(
                "Validating deserialized model failed: different keys"
            )
        for name, original_tensor in original_sd.items():
            tensorizer_tensor = tensorizer_sd[name]
            if not torch.equal(original_tensor, tensorizer_tensor):
                raise RuntimeError(
                    "Validating deserialized model failed:"
                    f" weights for tensor {name} don't match"
                )
```

Reply:
Acknowledged. Will add this to batch.